specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,58 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism which can tie into the lifecycle of flag evaluation. They
+operate similarly to middleware in many web frameworks.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.
```

Comment:
> flag type may not be possible in some languages. You may want to make this conditional, like some of the other sections: 

> The language type system differentiates between strings, numbers, booleans and structures.

Reply:
We would still know what type was requested, right?  Like they will have to call `getBooleanValue`, no?

Reply:
> We would still know what type was requested, right? Like they will have to call getBooleanValue, no?

"getBooleanValue" only exists in languages implementations that have types. A completely type-less language would just implement `getValue`

Reply:
Updated to be conditional.

Reply:
I'm still confused as to how we wouldn't have that information, even in a completely type-less language. Why would a type-less language just implement getValue? Wouldn't we still track the type as metadata somewhere? Probably a conversation to be had outside of this PR.

Reply:
There exists languages where there simply isn't rigorous concepts around data, even at runtime. There are not types - but just "bytes" that can be interpreted in multiple ways.

We will probably never deal with such languages, so I'm not married to keeping this requirement.